### ISSUE TITLE
Can't cancel coroutine/future using JSON-RPC client

### ISSUE THREAD

**Issue by gschmottlach-xse** [pengutronix/aiohttp-json-rpc#25]
When used within the context of an asyncio coroutine/future, it's impossible to cleanly "cancel" the tasks associated with the JsonRpcClient. In my application which uses this module, exceptions can occur when processing the *main* coroutine. This is illustrated below:

```
def run():
    """Entry point for console_scripts
    """
    # Register a signal handler to handle Ctrl-C
    signal.signal(signal.SIGINT, signal_handler)

    # Get the default loop
    loop = asyncio.get_event_loop()
    try:
        # Run the main program until it exits (or Ctrl-C is hit) 
        loop.run_until_complete(main(loop, sys.argv[1:]))
    except Exception as e: 
        print("Caught an error: {}".format(e))
    except SystemExit:
        # Catch the system exit and ignore
        pass
    finally:
        # We need to cancel any scheduled coroutines/tasks
        # that are still active in order to exit cleanly
        shutdown(loop)
```

The *shutdown()* routine is as follows:

```
def shutdown(loop):
    """Shutdown and cancel all pending coroutines/tasks."""
    tasks = asyncio.gather(*asyncio.Task.all_tasks(loop=loop),
                            loop=loop, return_exceptions=True)
    tasks.add_done_callback(lambda t: loop.stop())
    tasks.cancel()

    while not tasks.done() and not loop.is_closed():
        try:
            loop.run_forever()
        except asyncio.CancelledError:
            pass
    loop.close()
```

Basically, any pending task/future is cancelled and the loop stopped when all of these tasks are "done". The problem, unfortunately, is that ```loop.run_forever()``` never returns because the ```asyncio.CancelledError``` exception (raised by ```tasks.cancel()```) is being eaten inside of client.py (JsonRpcClient). Specifically, inside ```async def _handle_msgs(self)``` the

```
while not self._ws.close():
```

loop will not exit when the ```asyncio.CancelledError``` is triggered during the call to

```
raw_msg = await self._ws.receive()
```
All of the code inside the ```while``` loop is wrapped in a try/except which merely logs the exception. I'd suggest adding an additional

```
except asyncio.CancelledError:
    raise
```
To re-raise this specific exception so the ```while``` loop will be exited (or even ```break``` instead of ```raise``` from the ```while``` loop). If this isn't done the ```loop.forever()``` in a client's clean-up code will never exit. The only alternative is to explicitly close the associated *JsonRpcClient* web-socket prior to invoking a similar *shutdown()* function as shown above so the ```while``` loop exits normally. This may be difficult to do depending on the client code and whether this web-socket is easily accessible via the ```client.disconnect()``` call.

Properly handling a ```asyncio.CancelledError``` exception within a Python *future*/coroutine could really help clients safely and cleanly cancel their JSON-RPC clients. In this case, just re-raising this error will allow the event loop to exit properly.

Please consider making this change . . .  Thanks!

**Comment by gschmottlach-xse** [comment]
Is there any activity on this request? Seems like a fairly innocuous change which has a dramatic affect on the ability to cancel any pending requests and allow the client to cleanly exit.

**Comment by fscherf** [comment]
@gschmottlach-xse Yes! I am working on a solution atm. I'm sorry that it takes so much time.

**Comment by gschmottlach-xse** [comment]
Understood . . . no need to apologize. Just didn't want you to forget this one since it's difficult to clean up nicely without it. I am thankful that you're actively supporting these requests.

**Comment by fscherf** [comment]
I reproduced your problem but i had no good idea how to test this, without changing the pytest fixture and adding potential flaws to the test suite.
I ended up adding the code you recommended and tested it by hand. For this time it's good enough and i noted for myself to add a test in the future.

Thanks for reporting and your patience

**Comment by fscherf** [comment]
Your fix is released in v0.10
